integration_test/third_party_apps_data/applications/zookeeper/centos_rhel/install

```diff
@@ -0,0 +1,52 @@
+set -e
+
+sudo yum install -y curl java
+
+sudo mkdir -p /opt/zookeeper/stage
+sudo curl \
+    "https://dlcdn.apache.org/zookeeper/zookeeper-3.7.0/apache-zookeeper-3.7.0-bin.tar.gz" \
```

Comment:
So the reason we did this was this is what zookeeper recommends doing rather than a package manager. Their [getting started](https://zookeeper.apache.org/doc/current/zookeeperStarted.html). If you Google "install zookeeper" most tutorials follow the manual install route.

Reply:
++@qingling128, we should really be testing both but absent that it's preferable to use the distro packages when available.

Reply:
So I did some digging on this. The package manger installs either don't exist in base OS repositories or don't do much more than just downloading and unzipping. Debian/Ubuntu was the best one I could find and that still required creating users, finding and setting all the config files, creating a service file, and then running manually.

That being said I feel strongly we should leave the install scripts as is since this is likely how a customer will install. I had a hard time finding any online guide for installing Zookeeper that wasn't doing all of these manual steps in more or less the same order. I agree using a package manager is preferable but for Zookeeper it does not seem to be the way the community suggests installing and isn't documented very well.

Let me know your thoughts. I'm have a feeling doing package manager installs for centos and sles may raise the risk for flaky kokoro testing also as they don't seem to be well supported.

Reply:
```
root@unified-agents-dev-04:~# lsb_release -a
No LSB modules are available.
Distributor ID: Debian
Description:    Debian GNU/Linux 10 (buster)
Release:        10
Codename:       buster
root@unified-agents-dev-04:~# apt install zookeeper-bin zookeeperd
<snip>
root@unified-agents-dev-04:~# systemctl status zookeeper
● zookeeper.service - LSB: centralized coordination service
   Loaded: loaded (/etc/init.d/zookeeper; generated)
   Active: active (running) since Thu 2022-02-03 23:02:52 UTC; 13s ago
```

Zero configuration needed, I have a running zookeeper:

```
root@unified-agents-dev-04:~# /usr/share/zookeeper/bin/zkCli.sh stat /zookeeper
Connecting to localhost:2181

WATCHER::

WatchedEvent state:SyncConnected type:None path:null
cZxid = 0x0
ctime = Thu Jan 01 00:00:00 UTC 1970
mZxid = 0x0
mtime = Thu Jan 01 00:00:00 UTC 1970
pZxid = 0x0
cversion = -1
dataVersion = 0
aclVersion = 0
ephemeralOwner = 0x0
dataLength = 0
numChildren = 1
```

Reply:
Per sync offline, we can keep it as is in this PR. Will have separate conversation for the long-term follow up (Just shared a doc to discuss the overall strategy to test supported application versions.

Reply:
I would like to see @cpheps's response before we resolve this thread.

Reply:
@quentinmit What you did worked but curious how did you know what packages to install? Myself and another member of our team have spent time trying to figure out how to package manager install zookeeper and I found one article that listed the package just a pure `zookeeper` I've not see `zookeeper-bin` or `zookeeperd` in any article I came upon but maybe your searching skills are better than mine. 

Do you think it's a fair argument that most people will try `apt install zookeeper` see it doesn't fully work or search `how to install zookeeper` and find the manual installation steps? What you found is totally correct but I still think most customers will find guides leading them down the manual installation route.

Reply:
I just did `apt search zookeeper` on a Debian machine to find packages referencing zookeeper. I think that's a thing that most Debian sysadmins would do whenever looking for software to install. `zookeeper-bin` is actually not strictly necessary; it just provides more CLI tools so I could demonstrate the Zookeeper server is running. All you need to install to get a server is `zookeeperd`. This pattern of "common package containing base files, and separate packages for server and client" is super common on Debian distros, so I wasn't surprised by it. Usually the base package would have a name like "zookeeper-common", though.

You can also do a similar search at http://packages.debian.org/zookeeper if you prefer a webpage, though I would guess more users know about `apt search` than do the web page. I think "search for a package in my distribution" is a thing that most distribution users would do before looking for guides online.

Reply:
Ha, ok so I think we proved I'm not a linux master as I forgot about `apt search`, thanks for the info.

Can we say with confidence if a typical user would or would not do an `apt search` to find the package? If our goal for these installation scripts is to install the way a user would I feel there's a strong argument for the manual method. For Apache software in general the community seems to gravitate towards doing manual installs. I'm basing this on the fact that most web page or video tutorials all follow the manual installation path of getting it from Apache and doing the manual setup. If our goal is the least involved install then the package manager makes sense.

If we do switch to using `apt` for debian/ubuntu are we ok with the installation being fragmented where SLES and CentOS are still using the manual install? I did just verify that even doing the search in centos/sles still does not contain a zookeeper package.

Reply:
> Can we say with confidence if a typical user would or would not do an `apt search` to find the package? If our goal for these installation scripts is to install the way a user would I feel there's a strong argument for the manual method. For Apache software in general the community seems to gravitate towards doing manual installs. I'm basing this on the fact that most web page or video tutorials all follow the manual installation path of getting it from Apache and doing the manual setup. If our goal is the least involved install then the package manager makes sense.

I don't have any strong data for Zookeeper in particular one way or the other. I don't think you should necessarily extrapolate "Apache software" as a group, though - the Apache foundation sponsors a bunch of separate open source projects, and those projects don't really share anything in how they are released and deployed. For example, approximately zero users will install the Apache web server from anything other than distribution packaging. The only data I have for Zookeeper is Debian's [`popcon` data](https://qa.debian.org/popcon.php?package=zookeeper), which shows that Zookeeper is installed from packaging on 0.05% of Debian systems. (This is super noisy because 0.05% is only 103 machines; the vast majority of Debian users don't opt in to contributing install data to `popcon`.) I think the better grouping to apply might be "Java servers" - in general those are poorly or not at all packaged in distributions. Zookeeper definitely seems to be an exception to that general rule.

> If we do switch to using `apt` for debian/ubuntu are we ok with the installation being fragmented where SLES and CentOS are still using the manual install? I did just verify that even doing the search in centos/sles still does not contain a zookeeper package.

CentOS generally has a much smaller selection of software than other distributions, so we're going to find that a bunch of software is not packaged in CentOS. There's a second repo called "EPEL" (Extra Packages for Enterprise Linux) that has a bunch of packages that are in the default repos on other distros, but I don't see Zookeeper in EPEL either. EPEL is technically part of the Fedora project, so EPEL packages will show up in the search at https://packages.fedoraproject.org/search?query=zookeeper

SLES I think tends to have more software, though, and Zookeeper does appear to be "experimental" packaged in SLES 15: https://software.opensuse.org/package/zookeeper

We're getting into the realm of opinion now, but IMO it's fine to use different installation methods on different distributions, and in fact that's probably better for our integration test coverage. As long as one of the distributions is testing manual installation, I think we've covered that integration test case.

Reply:
@qingling128 Pinging for visibility on the proposed solution.

I think at this point it's going back and forth between @quentinmit's opinion and mine. There is a good point for if debian uses apt to install it'll give better coverage for integration tests. My only worry is we have a single config for all distros so if the apt package ever deviates from what we set in the other distros we could get failures. If we're comfortable with that risk then I'll switch debian to using apt.

@qingling128 @quentinmit Please let me know if you're ok with the above risk and want to move forward with changing debian to use `apt` install.

Reply:
Can you elaborate on "we have a single config for all distros"? Do you mean the Ops Agent config YAML we're testing with?

We should be testing defaults in that YAML, and if the defaults don't work on a particular distro, that's a legitimate test failure and we need to fix our defaults (or in very rare cases perhaps conditionalize the Ops Agent config, which I think we could do since it's still written by a shell script).

IMO if you can try it and the test works I would prefer to see that in this PR; if there's something other than "the defaults are incompatible" that blocks it, it's not important to improve in this PR and we can leave it as a TODO for future coverage improvement.

Reply:
My 2 cents: If `apt install` works as a drop-in replacement, let's switch to that. If there is additional setup to get it work, table it for now until we have an overall strategy decided for how to test the supported app versions. As for the pros and cons for the strategy moving forward, let's take the conversation to the `Test Ops Agent Supported Application Versions` doc instead. (Because once this PR gets merged, it's probably hard to try to find this thread).

Either way, it's not a blocker for this PR. We can do a follow up PR to swap to `apt install`, since swapping out the application setup logic in the tests is not a breaking change that we have to wait for.

Reply:
I played around with this a bit and the only issue I found that was decently different between the apache provided package and the debian package was the debian `log4j.properties` file is decently different. I will try to address this in the follow up logging PR to come. It looks like debian package manager even changed the default logging format.

Reply:
Sounds good. Thanks for checking. I'll leave this thread unresolved so that we can easily come back to this for reference when working on the logging PR (GitHub makes it extremely hard to locate a resolved thread).

But the action items are not a blocker for this current PR.

Reply:
@cpheps QQ to help decide whether we can release the metrics side in the coming release before the logging side lands:

You mentioned that `I played around with this a bit and the only issue I found that was decently different between the apache provided package and the debian package was the debian log4j.properties file is decently different.`. I assume d that means the metrics side still works and no user visible change of the implementation is needed. In other words the metrics side of the tests still pass even when we do `apt install`. Is that assumption right? If that's the case, I think we can release the metrics side first.

Reply:
Yeah, metrics should be agnostic on how it was installed. That's a logs receiver issue. I would say it's safe to release metrics.

Reply:
Got it. Thanks for confirming!